Neos.Neos/Classes/Security/Authorization/ContentRepositoryAuthorizationService.php

```diff
@@ -0,0 +1,200 @@
+use Neos\Flow\Security\Policy\PolicyService;
+use Neos\Flow\Security\Policy\Role;
+use Neos\Neos\Domain\Model\NodePermissions;
+use Neos\Neos\Domain\Model\User;
+use Neos\Neos\Domain\Model\WorkspacePermissions;
+use Neos\Neos\Domain\Model\WorkspaceRole;
+use Neos\Neos\Domain\Model\WorkspaceRoleSubject;
+use Neos\Neos\Domain\Model\WorkspaceRoleSubjects;
+use Neos\Neos\Domain\Service\WorkspaceService;
+use Neos\Neos\Security\Authorization\Privilege\EditNodePrivilege;
+use Neos\Neos\Security\Authorization\Privilege\ReadNodePrivilege;
+use Neos\Neos\Security\Authorization\Privilege\SubtreeTagPrivilegeSubject;
+use Neos\Party\Domain\Service\PartyService;
+
+/**
+ * Central point which does ContentRepository authorization decisions within Neos.
+ *
+ * @api
+ */
+#[Flow\Scope('singleton')]
+final readonly class ContentRepositoryAuthorizationService
+{
+    private const FLOW_ROLE_EVERYBODY = 'Neos.Flow:Everybody';
+    private const FLOW_ROLE_ANONYMOUS = 'Neos.Flow:Anonymous';
+    private const FLOW_ROLE_AUTHENTICATED_USER = 'Neos.Flow:AuthenticatedUser';
+    private const FLOW_ROLE_NEOS_ADMINISTRATOR = 'Neos.Neos:Administrator';
+
+    public function __construct(
+        private PartyService $partyService,
+        private WorkspaceService $workspaceService,
+        private PolicyService $policyService,
+        private PrivilegeManagerInterface $privilegeManager,
+    ) {
+    }
+
+    /**
+     * Determines the {@see WorkspacePermissions} an anonymous user has for the specified workspace (aka "public access")
+     */
+    public function getWorkspacePermissionsForAnonymousUser(ContentRepositoryId $contentRepositoryId, WorkspaceName $workspaceName): WorkspacePermissions
+    {
+        $subjects = [WorkspaceRoleSubject::createForGroup(self::FLOW_ROLE_EVERYBODY), WorkspaceRoleSubject::createForGroup(self::FLOW_ROLE_ANONYMOUS)];
+        $userWorkspaceRole = $this->workspaceService->getMostPrivilegedWorkspaceRoleForSubjects($contentRepositoryId, $workspaceName, WorkspaceRoleSubjects::fromArray($subjects));
+        if ($userWorkspaceRole === null) {
+            return WorkspacePermissions::none(sprintf('Anonymous user has no explicit role for workspace "%s"', $workspaceName->value));
+        }
+        return WorkspacePermissions::create(
+            read: $userWorkspaceRole->isAtLeast(WorkspaceRole::VIEWER),
+            write: $userWorkspaceRole->isAtLeast(WorkspaceRole::COLLABORATOR),
+            manage: $userWorkspaceRole->isAtLeast(WorkspaceRole::MANAGER),
+            reason: sprintf('Anonymous user has role "%s" for workspace "%s"', $userWorkspaceRole->value, $workspaceName->value),
+        );
+    }
+
+    /**
+     * Determines the {@see WorkspacePermissions} the given user has for the specified workspace
+     */
+    public function getWorkspacePermissionsForAccount(ContentRepositoryId $contentRepositoryId, WorkspaceName $workspaceName, Account $account): WorkspacePermissions
```

Comment:
> Do we want to keep the expandAccountRoles roles logic here

I would say it's an ok solution _for now_. We could centralize that logic to Flow in a separate PR, but that needs some more thought – especially if it adds runtime caches.
Also with the current API it would need to be sth like `PolicyService::expandRolesForAccount()` or similar (or we expand the `Account` model to already contain the expanded roles) because otherwise we would have to pass that untyped `$roles` array from the outside

> Also whenever we duplicate logic like neosUserFromAccount i would plead for adding documentation

The implementation is a one-liner basically:
```php
$this->partyService->getAssignedPartyOfAccount($account)
```
and that's just a core feature of the Party package. But I'm always in favor of more docs. Where would you suggest to add them and in what way?

Reply:
@mhsdesign can we mark this one resolved for now?

Reply:
this is more of a FYI for other reviewers so that can stay open for ever :)